BIP324: Cipher suite

dhruv opened the pull request:
This PR supersedes #20962 and introduces a two-layered cipher suite used in the latest draft of [BIP324](https://github.com/bitcoin/bips/pull/1378).

* Inner layer uses [RFC8439](https://datatracker.ietf.org/doc/html/rfc8439) which comes with a formal [security analysis](https://eprint.iacr.org/2014/613.pdf) so any novelty introduced by our cipher suite still offers a baseline confidence in confidentiality and authenticity. The RFC8439 instance is re-keyed every 256 messages for forward secrecy.
* Outer layer uses a forward secure version of ChaCha20, `FSChaCha20` which re-keys itself every 256 messages for forward secrecy. It is used to encrypt the message length resulting in a pseudorandom byte stream.

The dependency tree for BIP324 PRs is [here](https://bip324.com/sections/code-review/).

dhruv committed:
BIP324 Cipher Suite

dhruv committed:
RFC8439 nonce and counter for ChaCha20

dhruv committed:
RFC8439 implementation and tests

dhruv committed:
Adding forward secure FSChaCha20

dhruv commented:
Rebased

dhruv commented:
Found necessary fixes from sanitizers, fuzz tests and unit tests in downstream branches. Pushed those changes. Ready for further review.

dhruv commented on src/test/crypto_tests.cpp:
```diff
@@ -162,6 +164,34 @@ static void TestChaCha20(const std::string &hex_message, const std::string &hexk
+        std::vector<unsigned char> keystream;
+        keystream.resize(CHACHA20_ROUND_OUTPUT);
+        c20.Keystream(keystream.data(), CHACHA20_ROUND_OUTPUT);
+        BOOST_CHECK_EQUAL(HexStr(keystream).substr(0, hex_expected_keystream.size()), hex_expected_keystream);
```
No reason other than I'm using the test vectors as-is from the hyperlink in the comment before those vectors.

dhruv committed:
Allow for RFC8439 AD in cipher suite interface

dhruv commented:
Updated to:
- Remove multiple plaintext interface from RFC8439 -- it ended up not being super helpful to get to the optimization I was trying
- No longer expose the mac tag from RFC8439 - the tag is a detail internal to the AEAD and the ciphertext + mac is now treated as a blob

Ready for further review.

stratospher commented on src/crypto/chacha20.h:
```diff
@@ -39,4 +45,48 @@ class ChaCha20
+private:
+    ChaCha20 c20;
+    size_t rekey_interval;
+    uint32_t messages_with_key;
```
9d9ffc4: naming it as a counter would be more intuitive.
```suggestion
    uint32_t message_counter;
```

dhruv replied:
Fixed.

stratospher commented on src/crypto/bip324_suite.cpp:
```diff
@@ -0,0 +1,122 @@
+{
+    // check buffer boundaries
+    if (
+        // if we encrypt, make sure the destination has the space for the length field, header, ciphertext and MAC
```
361e9718:
```suggestion
        // if we encrypt, make sure the destination has the space for the length field, header, payload ciphertext and MAC
```

dhruv replied:
Updated.

stratospher commented on src/crypto/bip324_suite.cpp:
```diff
@@ -0,0 +1,122 @@
+        uint32_t ciphertext_len = BIP324_HEADER_LEN + input.size();
+        WriteLE32(reinterpret_cast<unsigned char*>(&ciphertext_len), ciphertext_len);
+
+        std::vector<std::byte> input_vec;
+        input_vec.resize(BIP324_HEADER_LEN + input.size());
```
361e971: `input_vec` can be defined directly with the required size.

dhruv replied:
Fixed.

dhruv commented on src/bench/rfc8439.cpp:
```diff
@@ -5,6 +5,7 @@
 #include <assert.h>
 #include <bench/bench.h>
 #include <crypto/rfc8439.h>
+#include <span.h>
```
Done where applicable.

dhruv commented on src/crypto/chacha20.h:
```diff
@@ -31,4 +45,48 @@ class ChaCha20
+        : c20{reinterpret_cast<const unsigned char*>(key.data()), FSCHACHA20_KEYLEN},
+          rekey_interval{rekey_interval},
+          messages_with_key{0},
+          rekey_counter{0},
```
Fixed.

dhruv commented on src/crypto/bip324_suite.h:
```diff
@@ -0,0 +1,78 @@
+static constexpr size_t REKEY_INTERVAL = 256;        // messages
+static constexpr size_t NONCE_LENGTH = 12;           // bytes
+
+enum BIP324HeaderFlags : uint8_t {
```
It's useful to have the implicit type conversion here for fuzz testing as the header flags are a byte that a peer could stuff with arbitrary bits.

dhruv commented on src/crypto/bip324_suite.h:
```diff
@@ -0,0 +1,78 @@
+    BIP324CipherSuite(const BIP324Key& K_L, const BIP324Key& K_P, const BIP324Key& rekey_salt)
+        : fsc20{K_L, rekey_salt, REKEY_INTERVAL},
+          rekey_ctr{0},
+          msg_ctr{0},
```
Fixed.

dhruv commented:
Latest push:
- Rebased
- Based on #26153 instead of #25354
- Addressed @stratospher comments
- Addressed @aureleoules comments
- Changed nomenclature to match latest BIP from working group

dhruv commented:
Latest push:
* Changes in working group:
  * Nomenclature changes
  * Encrypted packet length is now just length of contents rather than len(header + contents)
  * Rekey salt is 23 bytes instead of 32 bytes to allow just one SHA256 compression to achieve `new_key = SHA256(rekey_salt || old_key)`
  
Improvements:
* Cache mid-state when rekeying

Bug fixes:
* Found that previous code was using the little endian `contents_len` to perform encryption. Fixed.

dhruv commented:
Rebased. Bringing in upstream changes, ready for further review.

dhruv commented:
Rebased.

dhruv commented:
Rebased #26153

dhruv commented:
Latest push:
- Rebased upstream changes
- New rekey ratchet mechanism (forward security) based on ChaCha20 instead of SHA256. BIP changes are WIP and coming soon.

dhruv commented:
Rebased

dhruv commented:
Rebased changes from #26153

dhruv commented:
Rebased. Brought in changes from #26153.

dhruv commented:
Rebased.

theStack commented on src/crypto/bip324_suite.cpp:
```diff
@@ -0,0 +1,110 @@
+#include <cstring>
+#include <string.h>
+
+#ifndef HAVE_TIMINGSAFE_BCMP
+
+int timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
+{
+    const unsigned char *p1 = b1, *p2 = b2;
+    int ret = 0;
+
+    for (; n > 0; n--)
+        ret |= *p1++ ^ *p2++;
+    return (ret != 0);
+}
+
+#endif // TIMINGSAFE_BCMP
```
In commit 225023d9e3ea5a89037ef8a4f4404a0fdd3f1cf7: This function is not used anywhere in this file and hence can be removed (or rather moved to rfc8439.cpp, see other review comment).
```suggestion
```

dhruv replied:
I think I forgot to remove this previously used code. Thanks for pointing it out.

theStack commented on src/crypto/rfc8439.cpp:
```diff
@@ -0,0 +1,103 @@
+
+#include <cstring>
+
+#ifndef RFC8439_TIMINGSAFE_BCMP
+#define RFC8439_TIMINGSAFE_BCMP
+
+int rfc8439_timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
+{
+    const unsigned char *p1 = b1, *p2 = b2;
+    int ret = 0;
+
+    for (; n > 0; n--)
+        ret |= *p1++ ^ *p2++;
+    return (ret != 0);
+}
+
+#endif // RFC8439_TIMINGSAFE_BCMP
```
In commit 7a9d2fb8cf11eef8b6bd87ad1066df374bf3b620: Is there any reason to prefix the define and the function with "rfc8439"? Note that the point here is to let the build system detect if the function `timingsafe_bcmp` is available on the system (if yes, the define `HAVE_TIMINGSAFE_BCMP` is set) and only use the custom implementation if we need to. With this current construct, we would always use the custom one, since `RFC8439_TIMINGSAFE_BCMP` would obviously never be set. This should probably just be like on master (or in bip324_suite.cpp, see other review comment):
```suggestion
#ifndef HAVE_TIMINGSAFE_BCMP
#define HAVE_TIMINGSAFE_BCMP

int timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
{
    const unsigned char *p1 = b1, *p2 = b2;
    int ret = 0;

    for (; n > 0; n--)
        ret |= *p1++ ^ *p2++;
    return (ret != 0);
}

#endif // HAVE_TIMINGSAFE_BCMP
```

dhruv replied:
Done. Although I did leave it as-is in the second commit and addressed it in the fourth commit because I was getting a linker error for multiple definitions trying it on the second commit. The net change isn't different however.

fanquake replied:
> The net change isn't different however.

No, the net change is you've pointlessly broken things like git bisect. Please try and avoid doing that.

dhruv replied:
@fanquake not sure I understand. At which commit is the build broken? I specifically did it in a way that avoided that but may be misunderstanding something.

fanquake replied:
@dhruv ah, my misread was that you'd left the linker errors between commits 2 & 4. Although, I think introducing this, just to change in later commits is still not ideal. Will re-mark this as resolved, and post new review below.

theStack commented on src/crypto/rfc8439.cpp:
```diff
@@ -0,0 +1,103 @@
+}
+
+void ComputeRFC8439Tag(const std::array<std::byte, POLY1305_KEYLEN>& polykey,
+                       Span<const std::byte> aad, Span<const std::byte> ciphertext,
```
nit: const-correctness
```suggestion
                       const Span<const std::byte> aad, const Span<const std::byte> ciphertext,
```

theStack commented on src/crypto/rfc8439.cpp:
```diff
@@ -0,0 +1,103 @@
+    return polykey;
+}
+
+void RFC8439Crypt(ChaCha20& c20, const Span<const std::byte> in_bytes, Span<std::byte> out_bytes)
```
nit: seems like this function is never needed in another module, also not in follow-up PRs (checked at #24545)
```suggestion
static void RFC8439Crypt(ChaCha20& c20, const Span<const std::byte> in_bytes, Span<std::byte> out_bytes)
```

theStack reviewed (commented):
> Addressed review by @theStack.

Thanks! Another (non-blocking) suggestion: considering that the keys for RFC8439 have a fixed size, would it be worth it to introduce a `RFC8439Key` type that uses `std::array<std::byte, RFC8439_KEYLEN>`? I feel that's the slightly cleaner interface with enforcing the length already at compile-time, and we could remove the asserts in the `RFC8439Encrypt`/`RFC8439Decrypt` functions. The only annoying part is that the test/benchmark/fuzz code gets a bit longer, as one needs to convert from vector to array, and for the zero_key, one can't use the comfortable constructor for initializing the object with a repeated count of items.

fanquake commented on src/crypto/rfc8439.cpp:
```diff
@@ -0,0 +1,103 @@
+#ifndef RFC8439_TIMINGSAFE_BCMP
+#define RFC8439_TIMINGSAFE_BCMP
+
+int rfc8439_timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
```
In 7a9d2fb8cf11eef8b6bd87ad1066df374bf3b620:
It is fairly awkward to introduce a second copy of `timingsafe_bcmp` (one already in `chacha_poly_aead`), with a different name, and broken #ifdef logic, just to then rename it when you delete the original `timingsafe_bcmp` in a later commit. Although I guess re-using/extracting `timingsafe_bcmp` early from the to-be-deleted file is also a bit awkward.

fanquake commented on src/crypto/rfc8439.h:
```diff
@@ -0,0 +1,21 @@
+
+void RFC8439Encrypt(const Span<const std::byte> aad, const Span<const std::byte> key, const std::array<std::byte, 12>& nonce, const Span<const std::byte> plaintext, Span<std::byte> output);
+
+// returns false if authentication fails
```
In 7a9d2fb8cf11eef8b6bd87ad1066df374bf3b620: Make these Doxygen comments?

fanquake commented on src/crypto/rfc8439.h:
```diff
@@ -0,0 +1,22 @@
+#include <cstddef>
+#include <vector>
+
+constexpr static size_t RFC8439_KEYLEN = 32;
```
In 7a9d2fb8cf11eef8b6bd87ad1066df374bf3b620: These could be
```suggestion
constexpr static size_t RFC8439_KEYLEN{32};
```

fanquake commented on src/bench/rfc8439.cpp:
```diff
@@ -0,0 +1,76 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <assert.h>
```
In 7a9d2fb8cf11eef8b6bd87ad1066df374bf3b620: nit use c++ headers, i.e `<cassert>`

fanquake commented on src/bench/bip324_suite.cpp:
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019-2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
nit: additional newline

fanquake commented on src/bench/bip324_suite.cpp:
```diff
@@ -0,0 +1,117 @@
+#include <assert.h>
+#include <bench/bench.h>
+#include <crypto/bip324_suite.h>
+#include <crypto/rfc8439.h> // for the RFC8439_EXPANSION constant
```
Please don't add `for xyz` comments. They are unmaintainable, and if someone wants to know why an include is used, they can look at IWYU etc. Note that you could also add newly-added files to `ci/test/06-script_b.sh` to have the includes checked.

fanquake commented on src/crypto/bip324_suite.cpp:
```diff
@@ -0,0 +1,98 @@
+
+        std::vector<std::byte> header_and_contents(BIP324_HEADER_LEN + input.size());
+
+        memcpy(header_and_contents.data(), &flags, BIP324_HEADER_LEN);
```
```suggestion
        std::memcpy(header_and_contents.data(), &flags, BIP324_HEADER_LEN);
```
Same for memset etc.

fanquake commented:
Closing for now. This will be picked up again later. BIP324 review attention should be directed towards #27479 and https://github.com/bitcoin-core/secp256k1/pull/1129.
